### DISCUSSION TITLE
Finding error in computing error for stokes equation

### DISCUSSION THREAD

**Discussion by A123-bansal123** [firedrakeproject/firedrake#2834] [Firedrake support]
Hello everyone 

Below I attached a code of stokes equation with homogeneous boundary condition and exact solution is divergence preserving but It doesn't work properly. The error doesn't decreases as the mesh become finer and finer. Please help me in resolving this error.

```
from firedrake import *
N = 8
# Define mesh
M = UnitSquareMesh(N, N)
# Define Function Spaces
V = VectorFunctionSpace(M, "Lagrange", 2)
W = FunctionSpace(M, "Lagrange", 1)
R=  FunctionSpace(M, "Real",0)
Z = V * W *R
# Define Test and 
u, p ,c = TrialFunctions(Z)
v, q ,d = TestFunctions(Z)

f = Function(V)
x, y = SpatialCoordinate(M)
source = as_vector((4*pi*pi*sin(2*pi*x),-4*pi*pi*sin(2*pi*y)))
f.interpolate(source)
u_exact = Function(V)
velocity = as_vector((sin(2*pi*y)*(1-cos(2*pi*x)),sin(2*pi*x)*(cos(2*pi*y)-1)))
u_exact.interpolate(velocity)
p_exact = Function(W)
p_exact.interpolate((2*pi*(cos(2*pi*y)-cos(2*pi*x))))
sig_e=grad(u_exact)
a = (inner(grad(u), grad(v)) - inner(p, div(v)) - inner(div(u), q))*dx + p*d*dx + q*c*dx
L = inner(f, v) * dx
 
bc = DirichletBC(Z.sub(0), u_exact , "on_boundary")
z = Function(Z)
solve(a == L, z, bc)
u, p, c = z.subfunctions

error = dot(u-u_exact,u-u_exact)*dx 
E = sqrt(abs(assemble(error)))
print("error =", E)
```

**Comment by ksagiyam** [comment]
A quick inspection indicates that the source term should be:
```
source = as_vector((4 * pi * pi * (-2 * sin(2 * pi * y) * cos(2 * pi * x) - sin(2 * pi * x) + sin(2 * pi * y)),
                    4 * pi * pi * ( 2 * sin(2 * pi * x) * cos(2 * pi * y) - sin(2 * pi * x) + sin(2 * pi * y))))
```
if `u_exact` and `p_exact` are to be kept unchanged.

**Comment by A123-bansal123** [comment]
Sir, Thank you for your response. your suggestion is correct.
mesh      error of u in L2 norm     order of convergence
8*8               0.0031931                      -
16*16          0.00021056                3.921
32*32           1.334e-05                  3.98
64*64          8.371e-07                    3.99

Here, I am getting 4th order of convergence of u in L2 norm but as am approximating my u with polynomial of order 2 so, according to approximation theory, the order of convergence of u in L2 norm should be 3.

why this change occur here ?

**Comment by A123-bansal123** [comment]
Sir, Thank you for your response. your suggestion is correct.
mesh error of u in L2 norm order of convergence

```

mesh    error of u in L_2 norm    order of convergence
8*
*8                 0.0031931                               -16**16
  0.00021056                       3.921
32*
*32            1.334e-05                          3.9864**64
8.371e-07                            3.99

```

Here, I am getting 4th order of convergence of u in L2 norm but as am
approximating my u with polynomial of order 2 so, according to
approximation theory, the order of convergence of u in L2 norm should be 3.

Why does this change occur here ?
